mantidproject/mantidimaging
# Reverts shared array creation to python multiprocessing

I had some time over the weekend to consider the best way to fix the issues on IDAaaS caused by SharedArray. Namely https://github.com/mantidproject/mantidimaging/issues/745 and https://github.com/mantidproject/mantidimaging/issues/747.

Then I remembered that I changed it from `multiprocessing` to `SharedArray` as I suspected that the former was leaking the memory, but this turned out to be wrong after the change [1]. It was just never reverted as it took a bit of time to implement in the first place (I definitely fell into the [sunk cost fallacy](https://www.behavioraleconomics.com/resources/mini-encyclopedia-of-be/sunk-cost-fallacy/) here, and hopefully I learned my lesson!).

But now it's time for it to go, as it's caused a lot of issues for it's supposed "benefits" that we may never use.

This PR reverts the shared array creation to implementation from https://github.com/mantidproject/mantidimaging/commit/df4b3e2a6983d299c9ce71ff53efeceb575d0c97#diff-6b06bb41e81d5a3b1ccac6da5b0f4098583cf306a3194550caa1a169fe1182b4 (look at `parallel/utility.py`)

The core of this change is in the first commit https://github.com/mantidproject/mantidimaging/commit/24d5b31388e8beb45d3835e8dfa070d638d6ecb4. The rest is just cleanup to `free_memory` and all the workarounds added for deleting arrays.

Interestingly, Python itself is adding a built-in module, much like `SharedArray`, called [`shared_memory`](https://docs.python.org/3/library/multiprocessing.shared_memory.html) which provides a similar functionality to `SharedArray`. Perhaps if we ever require the machine global named arrays in the future, that should be the way to go.

### To test

Before starting make sure to open `htop` or another RAM monitoring tool - you'll need to keep an eye on it.

#### Load and kill
- Load a stack
- Rebin it to dims 9999, 9999
	- If the stack is too big, this will fail with Not enough memory warning. Then just use `4096, 4096`, or skip and just duplicate data until you are close to 100% memory
- Once you get to 100%, open operations window, open recon window, and open and **close** a stack comparison window (from workflow -> compare)
- Close some of the stacks - memory usage should go down
- Copy it again or just kill the process from `htop` or similar
	- Note: this must be SIGKILL to prevent any `atexit`s running, although this PR also removes that

#### Safe apply
- Load a stack
- Keep an eye on used RAM post load
- Apply remove outliers (or anything else) with Safe Apply
- The RAM usage should double
- Choose the original - it should fall back to the post-load value
- Redo and choose the new stack, it should also fall back


[1] The real culprit was `ImageItem`, and it has bitten us again - if you [don't call imageItem.close()](https://github.com/mantidproject/mantidimaging/blob/master/mantidimaging/gui/windows/stack_choice/view.py#L164-L165) the memory gets leaked, even if you close the widget.
